### PR TITLE
fix: enforce strict syntax for `@charset` in no-invalid-at-rules

### LINKED ISSUE #185: Rule Change: Enforce strict syntax for @charset
### What rule do you want to change?

no-invalid-at-rules

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```css
@charset 'iso-8859-15'; /* Invalid: single quotes */
@charset  "UTF-8";      /* Invalid: two spaces */
```

### What does the rule currently do for this code?

Currently, the rule only checks if the prelude is a `<string>`, so both `@charset 'iso-8859-15';` and `@charset  "UTF-8";` are not reported as errors.

### What will the rule do after it's changed?

After the change, the rule will report an error for any `@charset` that:

- Does not use double quotes for the encoding,
- Does not have exactly one space after `@charset`,
- Is not immediately terminated with a semicolon.

### Participation

- [x] I am willing to submit a pull request to implement this change.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by thecalamiity**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

To enforce strict syntax requirements for the `@charset` rule. This ensures that only valid `@charset` declarations are accepted.

#### What changes did you make? (Give an overview)

- Updated the no-invalid-at-rules rule to strictly validate `@charset` syntax. Now enforces:
  - The encoding must be enclosed in double quotes.
  - There must be exactly one space after `@charset`.
  - The rule must be immediately terminated with a semicolon.
- Added tests to cover valid and invalid `@charset` usages.
- Updated the documentation to note the strict requirements for `@charset`.

#### Related Issues

Fixes #185

### CODE REVIEW

**Comment by lumirlumir** [comment]
ping @thecalamiity

File: docs/rules/no-invalid-at-rules.md [hunk 1]
```diff
@@ -73,6 +72,29 @@ Examples of **incorrect** code:
 }
 ```
 
+Note on `@charset`: Although it begins with an `@` symbol, it is not an at-rule. It is a specific byte sequence of the following form:
+
+```css
+@charset "<charset>";
+```
+
+where `charset` is a [`<string>`](https://developer.mozilla.org/en-US/docs/Web/CSS/string) denoting the character encoding to be used. It must be the name of a web-safe character encoding defined in the [IANA-registry](https://www.iana.org/assignments/character-sets/character-sets.xhtml), and must be double-quoted, following exactly one space character (U+0020), and immediately terminated with a semicolon.
```
Comment (Tanujkanti4441) [inline docs/rules/no-invalid-at-rules.md]:
```suggestion
where `<charset>` is a [`<string>`](https://developer.mozilla.org/en-US/docs/Web/CSS/string) denoting the character encoding to be used. It must be the name of a web-safe character encoding defined in the [IANA-registry](https://www.iana.org/assignments/character-sets/character-sets.xhtml), and must be double-quoted, following exactly one space character (U+0020) after `@charset`, and immediately terminated with a semicolon.
```

File: tests/rules/no-invalid-at-rules.test.js [hunk 2]
```diff
@@ -293,5 +295,206 @@ ruleTester.run("no-invalid-at-rules", rule, {
...
+						prelude: "'UTF-8'",
+						expected: "<string>",
+					},
+					line: 1,
+					column: 10,
+					endLine: 1,
+					endColumn: 17,
+				},
+			],
+		},
```
Comment (Tanujkanti4441) [inline tests/rules/no-invalid-at-rules.test.js]: In the case of a syntax error in `@charset`, the error message `invalidPrelude` seems incorrect, such as in this test, the message would be:

```
Invalid prelude ''UTF-8'' found for at-rule '@charset'. Expected '<string>'.
```

Here it says, `<string>` is expected to be used, which is already used here, but with wrong quotes.

And in the case of  `@charset "UTF-8"`, where a semicolon is missing, it has the same error message.

What do you think about adding an error message only to report syntax errors?

Comment (thecalamiity) [reply tests/rules/no-invalid-at-rules.test.js]: Are you suggesting that instead of reporting `invalidPrelude`, we should report a single, more general error for all these cases? Something like: 
```
invalidCharsetSyntax: "Invalid @charset syntax. Expected: @charset \"<encoding-name>\"; (double quotes, exactly one space after @charset, and a semicolon at the end)."
```

Comment (Tanujkanti4441) [reply tests/rules/no-invalid-at-rules.test.js]: Yes, but i feel this error message seems a bit descriptive; instead, we can just suggest the correct syntax, like for - 
`@charset "UTF-8"` message can be something like this:

```
Invalid @charset syntax. Expected '@charset "UTF-8";'
```

Please feel free to choose the words that you find more accurate here.

Comment (thecalamiity) [reply tests/rules/no-invalid-at-rules.test.js]: Just updated the `@charset` error message - PTAL when you can.

File: src/rules/no-invalid-at-rules.js [hunk 3]
```diff
@@ -74,15 +85,91 @@ export default {
...
+		 * @param {AtrulePlain} node The node representing the rule.
+		 */
+		function validateCharsetRule(node) {
+			const { name, prelude, loc } = node;
+
+			const charsetNameLoc = {
+				start: loc.start,
+				end: {
+					line: loc.start.line,
+					column: loc.start.column + name.length + 1,
```
Comment (snitin315) [inline src/rules/no-invalid-at-rules.js:111]: Should be `end`?

```suggestion
				end: {
					line: loc.end.line,
					column: loc.end.column + name.length + 1,
```

File: src/rules/no-invalid-at-rules.js [hunk 4]
```diff
@@ -74,15 +85,91 @@ export default {
...
+					},
+				});
+				return;
+			}
+
+			const nodeText = sourceCode.getText(node);
+			if (!charsetPattern.test(nodeText)) {
+				const preludeText = sourceCode.getText(prelude);
+				const encoding =
+					preludeText.match(charsetEncodingPattern)?.[1] ?? "UTF-8";
```
Comment (snitin315) [inline src/rules/no-invalid-at-rules.js]: Why do we need `?? "UTF-8"`? In which case will we reach the right operand?

Comment (thecalamiity) [reply src/rules/no-invalid-at-rules.js]: When the encoding is missing, e.g., `@charset "";`

File: src/rules/no-invalid-at-rules.js (hunk 3 above)
Comment (thecalamiity) [reply src/rules/no-invalid-at-rules.js:111]: Using `start` is correct because we only want to highlight the at-rule name (`@charset`).

Comment (snitin315) [reply src/rules/no-invalid-at-rules.js:111]: Shouldn't we report the whole node?

```
@charse "utf-8"
^^^^^^^^^^^^^^^

@charset "";
^^^^^^^^^^^^
```

File: src/rules/no-invalid-at-rules.js (hunk 4 above)
Comment (snitin315) [reply src/rules/no-invalid-at-rules.js]: Can we add this test case? empty string.

File: src/rules/no-invalid-at-rules.js (hunk 3 above)
Comment (thecalamiity) [reply src/rules/no-invalid-at-rules.js:111]: We report each issue individually rather than flagging the entire at-rule in one go.
For example, if the name is invalid we report `unknownAtRule`, and if the prelude is missing we report `missingPrelude`, etc.

File: src/rules/no-invalid-at-rules.js (hunk 4 above)
Comment (thecalamiity) [reply src/rules/no-invalid-at-rules.js]: I’ve added a test for the empty string case and changed the behavior so we no longer autofix when the encoding is missing or whitespace-only. Previously we defaulted to "UTF-8", but I found this surprising since it guessed a value the user didn’t provide. Now it just reports the issue without autofixing. What do you think of this approach?

**Review by snitin315 — APPROVED** [review]
LGTM, thanks! Leaving it open for 2nd review

**Review by nzakas — APPROVED** [review]
LGTM. Thanks.
